### DISCUSSION TITLE
Regression issue with model_config on Pydantic's BaseModel in version >0.113.0

### DISCUSSION THREAD

**Discussion by vichannnnn** [fastapi/fastapi#12167] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel, ConfigDict, StringConstraints
from typing import Annotated
app = FastAPI()

class CustomBaseModel(BaseModel):
    model_config = ConfigDict(
        arbitrary_types_allowed=True, regex_engine="python-re", from_attributes=True
    )

valid_username = Annotated[
    str, StringConstraints(strip_whitespace=True, pattern="^[a-zA-Z0-9]{4," "20}$")
]
valid_password = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True, pattern=r"^(?=.*[A-Z])(?=.*\W)[^\s]{8,20}$"
    ),
]

class AccountRegister(CustomBaseModel):
    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "username": "hima",
                    "password": "Password123@",
                }
            ]
        },
    }

    username: valid_username  # type: ignore
    password: valid_password  # type: ignore

@app.post("/")
async def create_account(data: AccountRegister):
    return data

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app)
```

### Description

In FastAPI version 0.111.0, the CustomBaseModel that uses `regex_engine` is properly parsing the Request body for the regex. However, in FastAPI version 0.113.0 and 0.114.0, you get the following error with the same exact self-contained reproducible code above:

```
pydantic_core._pydantic_core.SchemaError: Error building "str" validator:
  SchemaError: regex parse error:
    ^(?=.*[A-Z])(?=.*\W)[^\s]{8,20}$
     ^^^
error: look-around, including look-ahead and look-behind, is not supported
```

This originally happens when you do not do any regex_engine config for your BaseModel and use it as it is, but in this case, even using it causes the same error for version above >0.113.0

### Operating System

macOS

### FastAPI Version

0.111.0, 0.113.0 and 0.114.0

### Pydantic Version

2.9.0

### Python Version

3.12.2

### Additional Context

For the self-contained code, you can replicate the issue by simply using the same code but switching the version between 0.113.0 or 0.114.0 and 0.111.0. It will parse the response properly in 0.111.0, but it will throw the abovementioned error with the same exact code in >0.113.0

**Comment by sinisaos** [comment, marked as answer]
@vichannnnn I tried your example using `re.compile` like this

```python
import re

valid_username = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True, pattern=re.compile(r"[a-zA-Z0-9]{4," "20}$")
    ),
]
valid_password = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True, pattern=re.compile(r"(?=.*[A-Z])(?=.*\W)[^\s]{8,20}$")
    ),
]
```
and it works. Can you try that?

**Reply by vichannnnn** [reply]
@sinisaos Thank you for your response, it worked! 

Seems like explicitly passing the `compile` function into the `pattern` parameter is needed after version 0.111.0 then!
